node_modules/oae-zencoder/lib/api.js

```diff
@@ -0,0 +1,137 @@
+ *
+ * @param  {Object}     notification    The notification data received from Zencoder
+ * @param  {Function}   callback        Standard callback function
+ * @param  {Object}     callback.err    An error that occurred, if any
+ */
+
+var encodingCompleted = module.exports.encodingCompleted = function(notification, callback) {
+    log().info('Received encoding completed notification from Zencoder');
+    var job = notification.job;
+
+    if (job.state === 'finished') {
+        var output = notification.output;
+        var zencoderId = output.id.toString();
+
+        ZencoderDAO.getContentRevisionId(zencoderId, function(err, contentRevision) {
+            if (err) {
+                return callback(err);
+            } else if (!contentRevision || !contentRevision.contentId || !contentRevision.revisionId) {
+                log().warn('No preview job exists for zencoderId %s', zencoderId);
+                return callback('Non-existing Zencoder ID was provided');
+            }
+
+            log().info('Encoding of video content with the contentId %s finished successfully.', contentRevision.contentId);
+
+            _createAndSavePreviews(contentRevision, output, function(err) {
+                if (err) {
+                    return callback(err);
+                }
+
+                return callback();
+            });
+        });
+    }
+
+    else if (job.state === 'failed') {
+        return callback({'err': notification, 'msg': 'Zencoder could not transcode this video for preview'});
+    }
+
+    else {
+        return callback({'err': notification, 'msg': 'Could not process notification from Zencoder'});
+    }
+};
+
+/**
+ * Create and save the previews
+ *
+ * @param  {Object}     contentRevision   An object containing content item and revision IDs
+ * @param  {Object}     output            The output information from Zencoder
+ * @param  {Function}   callback          Standard callback function
+ * @param  {Object}     callback.err      An error that occurred, if any
+ * @api private
+ */
+var _createAndSavePreviews = function(contentRevision, output, callback) {
+    var ctx = new PreviewContext(PreviewProcessorAPI.getConfiguration(), contentRevision.contentId, contentRevision.revisionId);
+
+    ctx.login(function(err) {
+        if (err) {
+            return callback(err);
+        }
+
+        ctx.getContentData(function(err) {
+            if (ctx.revision.previews.total > 0) {
+                return callback('Previews have already been created for this revision');
+            }
+
+            ctx.addPreviewMetadata('video', util.format('remote:%s', output.url));
+            ctx.addPreviewMetadata('key', output['pass_through']);
+
+            var thumbnail = _.findWhere(output.thumbnails, {label: 'base'});
+        
+            PreviewUtil.downloadRemoteFile(thumbnail.images[0].url, util.format('%s/base.png', ctx.baseDir), function(err, path) {
```

Comment:
Should we do any handling for cases where `output.thumbnails` might not have all of these elements we're expecting? (e.g., doesn't have a thumbnail with label = base, that thumbnail doesn't have an images array, the images array is empty, it doesn't have `url`, etc...

Also, to get the first item we normally do `_.first(...)` which is a bit safer to avoid exceptions.